### PR TITLE
LIG-6056 DINO ViT Benchmark Codebase

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
A continuation of @guarin 's work: https://github.com/lightly-ai/lightly/pull/1778/

See https://linear.app/lightly/issue/LIG-6056/dino-vit-benchmark for more details.

- Add DINO ViT code
- Modify DINO Loss
- Improve Main Script
  - Align CLI input params: `ckpt_path`, `precision`, `strategy`, and `float32_matmul_precision`
  - `persistent_workers=True` for pre-training
  - For ViT evals: add `eval_method` param to let user choose the linear/finetune evaluation flavor between SimCLR and MAE regardless of pretraining method
- Improve Eval Implementation
  - All:
      - use `print_rank_zero` everywhere
      - use `strategy` from CLI input
      - Refactor `LinearClassifier` and `FineTuneClassifier` to use a `BaseClassifier` as the parent class.
      - For ViT evals: add `eval_method` param to `linear_eval` and `finetune_eval`
  - Linear Classifier
      - Subclass `LinearClassifierMAE` from `LinearClassifier` for MAE eval only
      - Autofit `feature_dim`
      - `persistent_workers=True`
  - FineTune Classifier
      - Move `FineTuneClassifier` to `lightly.utils.benchmarking`
      - Subclass `FineTuneClassifierMAE` from `FineTuneClassifier` for MAE eval only
      - Autofit `feature_dim`
      - `persistent_workers=True`

### CODE REVIEW

File: benchmarks/imagenet/resnet50/finetune_eval.py [hunk 1]
```diff
@@ -5,36 +5,13 @@
...
-from lightly.utils.benchmarking import LinearClassifier, MetricCallback
+from lightly.utils.benchmarking import FinetuneClassifier, MetricCallback
 from lightly.utils.dist import print_rank_zero
-from lightly.utils.scheduler import CosineWarmupScheduler
-
-
-class FinetuneEvalClassifier(LinearClassifier):
-    def configure_optimizers(self):
-        parameters = list(self.classification_head.parameters())
-        parameters += self.model.parameters()
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/finetune_eval.py:22]: I believe the only difference between `FinetuneEvalClassifier` and `LinearClassifier` is the learning rate. I think it would be better to expose the learning rate as argument in `LinearClassifier.__init__` than to create a new `FinetuneClassifier` class in `utils/benchmarking`. Wdyt?

File: benchmarks/imagenet/resnet50/finetune_eval.py [hunk 2]
```diff
@@ -83,7 +61,7 @@ def finetune_eval(
         shuffle=True,
         num_workers=num_workers,
         drop_last=True,
-        persistent_workers=False,
+        persistent_workers=True,
     )
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/finetune_eval.py:65]: Why is this change needed?

File: benchmarks/imagenet/resnet50/finetune_eval.py [hunk 3]
```diff
@@ -117,13 +95,12 @@ def finetune_eval(
         ],
         logger=TensorBoardLogger(save_dir=str(log_dir), name="finetune_eval"),
         precision=precision,
-        strategy="ddp_find_unused_parameters_true",
-        num_sanity_val_steps=0,
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/finetune_eval.py:121]: Why remove `num_sanity_val_steps`?

File: benchmarks/imagenet/resnet50/finetune_eval.py [hunk 4]
```diff
@@ -117,13 +95,12 @@ def finetune_eval(
...
+        strategy=strategy,
     )
-    classifier = FinetuneEvalClassifier(
+    classifier = FinetuneClassifier(
         model=model,
         batch_size_per_device=batch_size_per_device,
-        feature_dim=2048,
+        feature_dim=model.online_classifier.feature_dim,
```
Comment (guarin) [inline benchmarks/imagenet/resnet50/finetune_eval.py:103]: Smart way to get the feature dim!

File: benchmarks/imagenet/vitb16/finetune_eval.py [hunk 5]
```diff
@@ -127,6 +128,7 @@ def configure_optimizers(self):
 
 def finetune_eval(
     model: Module,
+    method: str,
```
Comment (guarin) [inline benchmarks/imagenet/vitb16/finetune_eval.py:131]: Should we add a `finetune_method` argument instead of `method`? Then users can set `--finetune-method=mae`  if they want to run the MAE fine-tuning recipe independent of the pretraining method.

File: benchmarks/imagenet/vitb16/linear_eval.py [hunk 6]
```diff
@@ -65,6 +67,7 @@ def configure_optimizers(self):
 
 def linear_eval(
     model: Module,
+    method: str,
```
Comment (guarin) [inline benchmarks/imagenet/vitb16/linear_eval.py:70]: Same as for `finetune_eval`

File: lightly/loss/dino_loss.py [hunk 7]
```diff
@@ -99,6 +99,7 @@ def forward(
         teacher_out: List[Tensor],
         student_out: List[Tensor],
         epoch: int,
+        teacher_temp: Union[Tensor, None] = None,
```
Comment (guarin) [inline lightly/loss/dino_loss.py:102]: Let's allow `teacher_temp` to be a float.

File: lightly/loss/ibot_loss.py
Comment (guarin) [inline lightly/loss/ibot_loss.py:1]: iBOTLoss is only needed for DINOv2, let's merge these changes later

File: lightly/utils/scheduler.py
Comment (guarin) [inline lightly/utils/scheduler.py:1]: linear warmup is only needed for DINOv2, let's merge it later and include tests

**Review by guarin — CHANGES_REQUESTED** [review]
This looks great! Left some comments

File: benchmarks/imagenet/resnet50/finetune_eval.py (hunk 1 above)
Comment (guarin) [reply benchmarks/imagenet/resnet50/finetune_eval.py:22]: Sounds good. Actually, I believe we might want to split them in the future anyways. In the future we should extend the linear classifier to multiple heads with different learning rates. This allows to quickly grid search linear evaluation. This is not possible with the fine-tuning classifier.

File: benchmarks/imagenet/resnet50/finetune_eval.py (hunk 2 above)
Comment (guarin) [reply benchmarks/imagenet/resnet50/finetune_eval.py:65]: I wasn't aware that we use `True` in some cases. But it sounds good to me 👍🏼

File: benchmarks/imagenet/resnet50/finetune_eval.py (hunk 3 above)
Comment (guarin) [reply benchmarks/imagenet/resnet50/finetune_eval.py:121]: I believe there was a problem with the warmup schedule or validation metrics if the sanity check ran through. This is why we disabled it. Should have made a comment (maybe there is one in the PR that added it, you can check git blame).

I would keep it for now to avoid accidentally breaking anything.

**Review by guarin — APPROVED** [review]
LGTM! Feel free to make final changes, resolve comments and merge.

File: benchmarks/imagenet/vitb16/finetune_eval.py [hunk 8]
```diff
@@ -29,13 +32,18 @@ def __init__(
...
-        freeze_model: bool = False,
     ) -> None:
         super().__init__(
-            model, batch_size_per_device, feature_dim, num_classes, topk, freeze_model
+            model,
+            batch_size_per_device,
+            lr,
+            feature_dim,
+            num_classes,
+            topk,
```
Comment (guarin) [inline benchmarks/imagenet/vitb16/finetune_eval.py:46]: I would use keyword arguments here to make sure all values are correctly passed

File: lightly/utils/benchmarking/linear_classifier.py [hunk 9]
```diff
@@ -153,9 +156,77 @@ def configure_optimizers(  # type: ignore[override]
...
+        num_classes: int = 1000,
+        topk: Tuple[int, ...] = (1, 5),
+    ) -> None:
+        super().__init__(
+            model,
+            batch_size_per_device,
+            lr,
+            feature_dim,
+            num_classes,
+            topk,
```
Comment (guarin) [inline lightly/utils/benchmarking/linear_classifier.py:179]: Keyword args

File: lightly/utils/benchmarking/linear_classifier.py [hunk 10]
```diff
@@ -153,9 +156,77 @@ def configure_optimizers(  # type: ignore[override]
...
+    def forward(self, images: Tensor) -> Tensor:
+        # For linear evaluation, we want to freeze the feature extractor.
+        with torch.no_grad():
+            features = self.model.forward(images).flatten(start_dim=1)
+
+        output: Tensor = self.classification_head(features)
+
+        return output
+
+    def get_trainable_parameters(self) -> List[torch.nn.Parameter]:
```
Comment (guarin) [inline lightly/utils/benchmarking/linear_classifier.py:191]:
```suggestion
    def get_trainable_parameters(self) -> List[Parameter]:
```

We always import classes directly: 
```
from package import Class
```
 and import modules for functions
```
import torch
t = torch.tensor([1, 2, 3])
```

File: lightly/utils/benchmarking/linear_classifier.py [hunk 11]
```diff
@@ -153,9 +156,77 @@ def configure_optimizers(  # type: ignore[override]
...
+        lr: float = 0.05,
+        feature_dim: int = 2048,
+        num_classes: int = 1000,
+        topk: Tuple[int, ...] = (1, 5),
+    ) -> None:
+        super().__init__(
+            model,
+            batch_size_per_device,
+            lr,
+            feature_dim,
```
Comment (guarin) [inline lightly/utils/benchmarking/linear_classifier.py:214]: Keyword args

File: lightly/utils/benchmarking/linear_classifier.py [hunk 12]
```diff
@@ -153,9 +156,77 @@ def configure_optimizers(  # type: ignore[override]
...
+    def forward(self, images: Tensor) -> Tensor:
+        # For finetuning, we want to update the feature extractor.
+        features = self.model.forward(images).flatten(start_dim=1)
+
+        output: Tensor = self.classification_head(features)
+
+        return output
+
+    # Type ignore is needed because return type of LightningModule.configure_optimizers
+    def get_trainable_parameters(self) -> List[torch.nn.Parameter]:
```
Comment (guarin) [inline lightly/utils/benchmarking/linear_classifier.py:228]:
```suggestion
    def get_trainable_parameters(self) -> List[Parameter]:
```

**Review by guarin — APPROVED** [review]
LGTM! Feel free to resolve and merge.

File: lightly/utils/benchmarking/linear_classifier.py [hunk 13]
```diff
@@ -151,9 +156,77 @@ def configure_optimizers(  # type: ignore[override]
...
+    def forward(self, images: Tensor) -> Tensor:
+        # For finetuning, we want to update the feature extractor.
+        features = self.model(images).flatten(start_dim=1)
+
+        output: Tensor = self.classification_head(features)
+
+        return output
+
+    # Type ignore is needed because return type of LightningModule.configure_optimizers
+    def get_trainable_parameters(self) -> List[Parameter]:
```
Comment (guarin) [inline lightly/utils/benchmarking/linear_classifier.py:228]: There is no type ignore :)

**Review by guarin — APPROVED** [review]
LGTM! Feel free to resolve and merge
